### PR TITLE
Move the sandbox application to React. Include Library Editor in repository.

### PR DESCRIPTION
**Pull request by dompuiu**
## Description

* Integrating the code editor inside the sandbox
* Move Lib Sandbox and View Sandbox to a React App
* Use @react-spectrum components

## Related Issue

These changes are required so I can start working on: https://jira.corp.adobe.com/browse/CORE-48967

## How Has This Been Tested?

Tested in a few local extensions I had.

## Types of changes

- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [x] Breaking change (fix or feature that would cause existing functionality to change)

## Checklist:

- [x] I have signed the [Adobe Open Source CLA](http://opensource.adobe.com/cla.html).
- [x] My code follows the code style of this project.
- [x] My change requires a change to the documentation.
- [x] I have updated the documentation accordingly.
- [x] I have read the **CONTRIBUTING** document.
- [x] I have added tests to cover my changes.
- [x] All new and existing tests passed.

### CODE REVIEW

**Comment by Aaronius** [comment]
It would be great if we could get these warnings to go away, since they show up for the extension developer running the sandbox (at least when I've npm linked it--maybe they don't show up for a prod build of the sandbox):

image

**Comment by dompuiu** [comment]
> It would be great if we could get these warnings to go away, since they show up for the extension developer running the sandbox (at least when I've npm linked it--maybe they don't show up for a prod build of the sandbox):
> 
> image

Those are triggered by some react-spectrum components. They are shown only when you link sandbox. They are shown by the developer server.

**Comment by Aaronius** [comment]
Okay, thanks. I figured that might be the case.

Here's something else. If I add a `debugger` statement somewhere in the react app in the sandbox, the app won't render because of:

image

That's kind of the pits when trying to debug.

**Comment by Aaronius** [comment]
I get this frequently when switching between views: 

image

You can reproduce it if you clone https://github.com/adobe/reactor-extension-alloy, run `npm ci`, then run `npm run dev` and try to switch between views in the view sandbox.

**Comment by dompuiu** [comment]
> I get this frequently when switching between views:
> 
> image
> 
> You can reproduce it if you clone https://github.com/adobe/reactor-extension-alloy, run `npm ci`, then run `npm run dev` and try to switch between views in the view sandbox.

Yes, I know. It's annoying. The react-spectrum fixed it:
https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/utils/src/useResizeObserver.ts#L16-L44

Hopefully they will release it soon.

**Comment by dompuiu** [comment]
> Okay, thanks. I figured that might be the case.
> 
> Here's something else. If I add a `debugger` statement somewhere in the react app in the sandbox, the app won't render because of:
> 
> image
> 
> That's kind of the pits when trying to debug.
> 
> Edit: When trying to debug the sandbox, that is.

I know. I got used to the rule. I find the rule useful because it minimizes the risk to forget a debugger somewhere in the code. In VSCode I have a shortcut where I can disable the warning for the line where I add the debugger. Probably there is something similar in WebStorm. Alternatively, you can change the ESLint rules if you want.

**Comment by Aaronius** [comment]
I notice when we're using the lib API to run the sandbox:

```
const sandbox = require("@adobe/reactor-sandbox");
sandbox.run();
```

and then our process finishes, the sandbox still seems to hold onto port 3001. I would guess this is because of how the sandbox is spawning the react app process in executeSandboxComponents.js.

You can reproduce it if you clone https://github.com/adobe/reactor-extension-alloy, run `npm ci,` open up `test/functional/actions/resetEventMergeId.spec.js` and change `test(` to `test.only(` (otherwise the test run will take a long time), then run `npm run test:functional`. After the process finishes, run `npm run test:functional` again, and you'll see:

image

**Review by Aaronius — COMMENTED** [review]
> > I get this frequently when switching between views:
> > image
> > You can reproduce it if you clone https://github.com/adobe/reactor-extension-alloy, run `npm ci`, then run `npm run dev` and try to switch between views in the view sandbox.
> 
> Yes, I know. It's annoying. The react-spectrum fixed it:
> https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/utils/src/useResizeObserver.ts#L16-L44
> 
> Hopefully they will release it soon. I patched my node modules file, so I don't see it anymore.

I wouldn't recommend releasing unless you can incorporate the patch. The sandbox is basically unusable for us as it is.

**Comment by Aaronius** [comment]
I scanned the changes lightly, but then tested out the integration with our extension more in depth. I marked a couple comments with the 👀  emoji. Those are the comments that I think really need to get resolved before releasing. Otherwise, everything seems to be working well.

File: src/tasks/run.js [hunk 1]
```diff
@@ -65,98 +86,105 @@ const configureApp = app => {
...
+    } else {
+      res.sendFile(`${files.INIT_FILES_SRC_PATH}/${files.LIB_SANDBOX_HTML_FILENAME}`);
+    }
   });
 
-  app.get('/editor-container.js', function(req, res) {
+  // Redirect old /viewSandbox.html to the new path.
+  app.get(`/${files.VIEW_SANDBOX_HTML_FILENAME}`, (_, res) => {
+    res.redirect(301, '/#/viewSandbox');
+  });
```
[the remark this answers is not in the document]
Comment (dompuiu) [reply src/tasks/run.js:178]: The port 3001 is running only when you link the sandbox. You will do this only when you develop a new feature on sandbox. Once we build the sandbox, the result application will run also on port 3000. Given all of this, I can remove it, if you think it's useless.

File: src/tasks/run.js [hunk 2]
```diff
@@ -192,33 +233,25 @@ const configureApp = app => {
...
       if (error) {
         reject(error);
       } else {
         console.log(
-          '\nExtension sandbox running at http://localhost:' +
-            PORT +
-            ' and at https://localhost:' +
-            SSL_PORT
+          `\nExtension sandbox running at \
+http://localhost:${PORT} and at https://localhost:${SSL_PORT}`
```
[the remark this answers is not in the document]
Comment (dompuiu) [reply src/tasks/run.js:254]: The port 3001 is only running when you link the sandbox. The end result will only work on 3000 and 4000 like it previously did.

File: README.md [hunk 3]
```diff
@@ -41,8 +37,9 @@ npx @adobe/reactor-sandbox init
...
+  * `libSandbox.html` includes some simple HTML along with a single script tag called `launch-EN00000000000000000000000000000000.js`, which simulates a Production Launch Library. The simulated library contains the combined contents of your `container.js` file and the `Turbine` engine.
+  * `libSandbox.html` can be modified to manually test whatever you would like. For example, if you'd like to test that your brand new "focus" event type is working as desired, you can add a text input to the `libSandbox.html` to ensure your dummy rule fires when a form element receives focus.
```
[the remark this answers is not in the document]
Comment (dompuiu) [reply README.md:42]: This was written by Brent. I was thinking that he wrote it after he consulted with you.

**Comment by dompuiu** [comment]
> > > I get this frequently when switching between views:
> > > image
> > > You can reproduce it if you clone https://github.com/adobe/reactor-extension-alloy, run `npm ci`, then run `npm run dev` and try to switch between views in the view sandbox.
> > 
> > 
> > Yes, I know. It's annoying. The react-spectrum fixed it:
> > https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/utils/src/useResizeObserver.ts#L16-L44
> > Hopefully they will release it soon. I patched my node modules file, so I don't see it anymore.
> 
> I wouldn't recommend releasing unless you can incorporate the patch. The sandbox is basically unusable for us as it is.

I think if we merge this in master, we don't need to release a new version until they release a new version of react-spectrum. But merging it in master it would allow me to start working on adding SSF functionality.

File: src/cli.js [hunk 4]
```diff
@@ -26,11 +30,14 @@ switch (task) {
     execute = require('./tasks/init');
     break;
   default:
+    // eslint-disable-next-line no-console
+    console.log('IN RUN TASK ', path.join(__dirname));
```
[the remark this answers is not in the document]
Comment (dompuiu) [reply src/cli.js:34]: No. I think Brent added that for helping him understand the flow better.

**Comment by dompuiu** [comment]
> I notice when we're using the lib API to run the sandbox:
> 
> ```
> const sandbox = require("@adobe/reactor-sandbox");
> sandbox.run();
> ```
> 
> and then our process finishes, the sandbox still seems to hold onto port 3001. I would guess this is because of how the sandbox is spawning the react app process in executeSandboxComponents.js.
> 
> You can reproduce it if you:
> 
> 1. Clone https://github.com/adobe/reactor-extension-alloy
> 2. Run `npm ci,` open up `test/functional/actions/resetEventMergeId.spec.js` and change `test(` to `test.only(` (otherwise the test run will take a long time). Also change `http://localhost:3000/viewSandbox.html` to `http://localhost:3001/#/viewSandbox`.
> 3. Open `test/functional/helpers/switchToIframe.js` and change `#extensionViewIframe` to `iframe`
> 4. Run `npm run test:functional`.
> 5. After the process finishes, run `npm run test:functional` again, and you'll see:
> 
> image

Would it be possible to do a `npm pack`, install the archive and test to see if you have the same issue? The port 3001 won't be there on a built sandbox.

File: src/tasks/run.js (hunk 1 above)
[the remark this answers is not in the document]
Comment (Aaronius) [reply src/tasks/run.js:178]: I see. I didn't realize that. Whatever you want to do here is fine with me.

**Comment by Aaronius** [comment]
> > > > I get this frequently when switching between views:
> > > > image
> > > > You can reproduce it if you clone https://github.com/adobe/reactor-extension-alloy, run `npm ci`, then run `npm run dev` and try to switch between views in the view sandbox.
> > > 
> > > 
> > > Yes, I know. It's annoying. The react-spectrum fixed it:
> > > https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/utils/src/useResizeObserver.ts#L16-L44
> > > Hopefully they will release it soon. I patched my node modules file, so I don't see it anymore.
> > 
> > 
> > I wouldn't recommend releasing unless you can incorporate the patch. The sandbox is basically unusable for us as it is.
> 
> I think if we merge this in master, we don't need to release a new version until they release a new version of react-spectrum. But merging it in master it would allow me to start working on adding SSF functionality.

Okay, sounds good.

File: src/app/viewSandbox/components/ValidateTabContent.jsx [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+            ajv.addMetaSchema(require('ajv/lib/refs/json-schema-draft-04.json'));
+
+            return ajv.compileAsync(descriptor.schema).then((validate) => {
+              const matchesSchema = validate(settings);
+
+              if (matchesSchema) {
+                setValidationState(true);
+              } else {
+                setValidationState(
+                  'Settings object does not match schema.' +
```
Comment (Aaronius) [inline src/app/viewSandbox/components/ValidateTabContent.jsx:49]: Space after period.

**Comment by Aaronius** [comment]
I don't know what's causing this, but as I switch between the Init and Get Settings tabs I see an error icon show up in CodeMirror for a second or so.

**Comment by Aaronius** [comment]
It seems that views are sometimes initialized with an empty `initInfo` object, which isn't expected and extension developers like us haven't coded around inside the extension view. This results in errors being thrown from the extension view. I don't think we should be initializing the view with an empty `initInfo` object.

**Comment by Aaronius** [comment]
> > I notice when we're using the lib API to run the sandbox:
> > ```
> > const sandbox = require("@adobe/reactor-sandbox");
> > sandbox.run();
> > ```
> > 
> > 
> > and then our process finishes, the sandbox still seems to hold onto port 3001. I would guess this is because of how the sandbox is spawning the react app process in executeSandboxComponents.js.
> > You can reproduce it if you:
> > 
> > 1. Clone https://github.com/adobe/reactor-extension-alloy
> > 2. Run `npm ci,` open up `test/functional/actions/resetEventMergeId.spec.js` and change `test(` to `test.only(` (otherwise the test run will take a long time). Also change `http://localhost:3000/viewSandbox.html` to `http://localhost:3001/#/viewSandbox`.
> > 3. Open `test/functional/helpers/switchToIframe.js` and change `#extensionViewIframe` to `iframe`
> > 4. Run `npm run test:functional`.
> > 5. After the process finishes, run `npm run test:functional` again, and you'll see:
> > 
> > image
> 
> Would it be possible to do a `npm pack`, install the archive and test to see if you have the same issue? The port 3001 won't be there on a built sandbox.

This seems to fix the issue.

**Comment by Aaronius** [comment]
When I switch to the View Sandbox from the Library Sandbox, I see this briefly. Not a big deal, but kind of annoying.

image

**Comment by dompuiu** [comment]
> I don't know what's causing this, but as I switch between the Init and Get Settings tabs I see an error icon show up in CodeMirror for a second or so.

The code mirror is initialized with an empty string, which is not a valid JSON. At the next render it receives the init object. That is the cause.

**Comment by Aaronius** [comment]
> > > > > I get this frequently when switching between views:
> > > > > image
> > > > > You can reproduce it if you clone https://github.com/adobe/reactor-extension-alloy, run `npm ci`, then run `npm run dev` and try to switch between views in the view sandbox.
> > > > 
> > > > 
> > > > Yes, I know. It's annoying. The react-spectrum fixed it:
> > > > https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/utils/src/useResizeObserver.ts#L16-L44
> > > > Hopefully they will release it soon. I patched my node modules file, so I don't see it anymore.
> > > 
> > > 
> > > I wouldn't recommend releasing unless you can incorporate the patch. The sandbox is basically unusable for us as it is.
> > 
> > 
> > I think if we merge this in master, we don't need to release a new version until they release a new version of react-spectrum. But merging it in master it would allow me to start working on adding SSF functionality.
> 
> Okay, sounds good.

After I install the sandbox after using `npm pack`, I don't see this error anymore, I assume because the build process removing some of the debugging stuff from React. Maybe it doesn't need to hold up a release.

**Comment by dompuiu** [comment]
> > > > > > I get this frequently when switching between views:
> > > > > > image
> > > > > > You can reproduce it if you clone https://github.com/adobe/reactor-extension-alloy, run `npm ci`, then run `npm run dev` and try to switch between views in the view sandbox.
> > > > > 
> > > > > 
> > > > > Yes, I know. It's annoying. The react-spectrum fixed it:
> > > > > https://github.com/adobe/react-spectrum/blob/main/packages/%40react-aria/utils/src/useResizeObserver.ts#L16-L44
> > > > > Hopefully they will release it soon. I patched my node modules file, so I don't see it anymore.
> > > > 
> > > > 
> > > > I wouldn't recommend releasing unless you can incorporate the patch. The sandbox is basically unusable for us as it is.
> > > 
> > > 
> > > I think if we merge this in master, we don't need to release a new version until they release a new version of react-spectrum. But merging it in master it would allow me to start working on adding SSF functionality.
> > 
> > 
> > Okay, sounds good.
> 
> After I install the sandbox after using `npm pack`, I don't see this error anymore, I assume because the build process removing some of the debugging stuff from React. Maybe it doesn't need to hold up a release.

Those errors are shown by the dev server. That is why you can also close them. I think the error still is visible in the console.

**Comment by Aaronius** [comment]
> > I don't know what's causing this, but as I switch between the Init and Get Settings tabs I see an error icon show up in CodeMirror for a second or so.
> 
> The code mirror is initialized with an empty string, which is not a valid JSON. At the next render it receives the init object. That is the cause.

Can we fix it by not initializing it with an empty string or something like that? I'm not sure what we were doing before, but I don't remember seeing that in prior versions.

**Comment by dompuiu** [comment]
> Can we fix it by not initializing it with an empty string or something like that? I'm not sure what we were doing before, but I don't remember seeing that in prior versions.

I'll see what I can do. I might not render the code mirror until there is some content to show.

**Comment by dompuiu** [comment]
> When I switch to the View Sandbox from the Library Sandbox, I see this briefly. Not a big deal, but kind of annoying.
> 
> image

That is probably when the iframe is not rendered. I'll if I can do something to fix that.

**Comment by Aaronius** [comment]
> > After I install the sandbox after using `npm pack`, I don't see this error anymore, I assume because the build process removing some of the debugging stuff from React. Maybe it doesn't need to hold up a release.
> 
> Those errors are shown by the dev server. That is why you can also close them. I think the error still is visible in the console for the built sandbox.

I haven't seen the errors in the console, but they might be there somewhere. If you want, you could use https://github.com/ds300/patch-package to patch react-spectrum before releasing.

File: README.md [hunk 6]
```diff
@@ -41,8 +37,9 @@ npx @adobe/reactor-sandbox init
...
+  * `container.js` When Launch publishes a library, it consists of two parts: (1) a data structure that stores information about saved rules, data elements, and extension configuration and (2) an engine to operate on such a data structure. `container.js` is the data structure (not the engine) that you may modify to simulate saved rules, data elements, and extension configuration. This template will be used to produce a complete `container.js` (JavaScript from your extension will be added) which will be used in tandem with the `Turbine` engine inside the sandbox. We have tried to provide inline comments on how to modify `container.js`, but if you need further help, please [let the Launch team know](https://developer.adobelaunch.com/guides/extensions/development-resources/) and we can help you out.
+  * `libSandbox.html` includes some simple HTML along with a single script tag called `launch-EN00000000000000000000000000000000.js`, which simulates a Production Launch Library. The simulated library contains the combined contents of your `container.js` file and the `Turbine` engine.
```
[the remark this answers is not in the document]
Comment (dompuiu) [reply README.md:41]: @Aaronius Are you referring to the filename(`launch-EN00000000000000000000000000000000.js`)?

[the remark this answers is not in the document]
Comment (Aaronius) [reply README.md]: No, I mean "Production Launch Library" should just be "production Launch library".

**Comment by dompuiu** [comment]
> When I switch to the View Sandbox from the Library Sandbox, I see this briefly. Not a big deal, but kind of annoying.
> 
> image

I am not able to replicate this issue. Do you have any steps?

**Comment by dompuiu** [comment]
> It seems that views are sometimes initialized with an empty `initInfo` object, which isn't expected and extension developers like us haven't coded around inside the extension view. This results in errors being thrown from the extension view. I don't think we should be initializing the view with an empty `initInfo` object.

Is there any extension where I can see this error? From what I see, the extension bridge has an empty object as a default value.

**Comment by dompuiu** [comment]
Talk with @Aaronius in person. Except the Code Mirror error marker that is seen when the component is initialized, all the other issues were fixed. I'll merge this.

**Pull request merged**